scripts/check_samples.py

```diff
@@ -0,0 +1,121 @@
+import yaml
+
+def build_sample_set():
+    """
+    build_sample_set returns a set of sample kinds read from the samples/ folder
+
+    Raises:
+        yaml.YAMLError if any sample file is not a valid YAML file
+    Returns:
+        set of sample kind (string) to use in check_template_sample
+    """
+
+    #default sample_set to empty
+    sample_set = set()
+
+    # Retrieve the samples constraints in the samples folder
+    for sample_file_name in glob.glob("samples/*.yaml"):
+        with open(sample_file_name, 'r') as sample_file:
+            try:
+                sample_object = yaml.safe_load(sample_file)
+                sample_set.add(sample_object["kind"])
+            except yaml.YAMLError as error:
+                print("Error parsing sample {}: {}".format(sample_file, error))
+                sys.exit(1)
+    return sample_set
+
+def check_template_sample(template_file_name, sample_set):
+    """
+    check_template_sample checks if a template has a sample associated in the samples/ folder
+
+    Args:
+        template_file_name: the template file path to check
+        sample_set: set of sample kinds (string) built by build_sample_set 
+    Raises:
+        yaml.YAMLError if the input template or any sample file is not a valid YAML file
+    Returns:
+        Boolean - True if a sample was found for this template, False otherwise.
+    """
+
+    # retrieve the template kind
+    with open(template_file_name, 'r') as template_file:
+        try:
+            template_object = yaml.safe_load(template_file)
+            template_kind = template_object["spec"]["crd"]["spec"]["names"]["kind"]
+            sample_found = False
+
+            for sample_kind in sample_set:
+                if sample_kind == template_kind:
+                   sample_found = True
+                   break
+
+            # if not, error out
+            if not sample_found:
+                print("No sample found for template {} ({})".format(template_file_name, template_kind))
+            
+            return sample_found
+
+        except yaml.YAMLError as error:
+            print("Error parsing template {}: {}".format(template_file_name, error))
+            sys.exit(1)
+
+def check_template_samples(sample_set):
+    """
+    check_template_samples runs check_template_sample on all templates in policies/
+    
+    Args:
+        sample_set: set of sample kinds (string) built by build_sample_set 
+    """
+    
+    # Default missing_sample to False
+    missing_sample = False
+
+    print("Verifying sample files for all templates...")
+
+    # Reccurisvely look for templates in the policies/ folders
+    for template_file_name in glob.glob("policies/**/*.yaml", recursive=True):
+        
+        # excluding legacy templates
+        if not template_file_name.startswith("policies/templates/legacy"):
+
+            # only run the check_template_sample function on actual template
+            with open(template_file_name, 'r') as template_file:
```

Comment:
Opening the file twice isn't very efficient, could we consolidate this call into the check_template_sample (or pass in the loaded yaml)?

Reply:
Fixed - I passed the loaded yaml instead of the filename